test/shared/src/main/scala/zio/test/TestAspect.scala

```diff
@@ -199,14 +215,18 @@ object TestAspect {
   /**
    * An aspect that times out tests using the specified duration.
    */
-  def timeout(duration: Duration): TestAspect[Nothing, Clock, Nothing, Any] =
-    new TestAspect.PerTest[Nothing, Clock, Nothing, Any] {
-      def perTest[R >: Nothing <: Clock, E >: Nothing <: Any](test: ZIO[R, E, TestResult]): ZIO[R, E, TestResult] =
-        test.timeout(duration).map {
-          case None =>
-            Assertion
-              .failure(FailureDetails.Runtime(Cause.fail(new TimeoutException(s"Timeout of ${duration} exceeded"))))
-          case Some(v) => v
+  def timeout(duration: Duration): TestAspect[Nothing, Live[Clock], Nothing, Any] =
+    new TestAspect.PerTest[Nothing, Live[Clock], Nothing, Any] {
+      def perTest[R >: Nothing <: Live[Clock], E >: Nothing <: Any](
+        test: ZIO[R, E, TestResult]
+      ): ZIO[R, E, TestResult] =
+        ZIO.environment[R].flatMap { r =>
+          Live.live(test.provide(r).timeout(duration)).map {
```

Comment:
This is really interesting (get the environment, provide it on the inside, do an outside effect, lift that up to `Live`). 

It makes me think we can have a combinator:

```scala
def mockLive[R, R1, E, A](zio: ZIO[R, E, A])(f: IO[E, A] => ZIO[R with Live[R1], E, B]): ZIO[R, E, B] = 
  ZIO.environment[R].flatMap(r => Live.live(f(zio.provide(r))))
```

Then use it like so:

```scala
Live.mockLive(test, _timeout(duration))
```

I don't like the name and there may be an even more generic combinator for this....

Reply:
Maybe something like?

```scala
def withLive[R, R1, E, A, B](zio: ZIO[R, E, A])(f: IO[E, A] => ZIO[R1, E, B]): ZIO[R with Live[R1], E, B] =
  ZIO.environment[R].flatMap(r => live(f(zio.provide(r))))
```

The type inference on the second parameter seems to be pretty poor. I don't know if there is anything we can do about that.

Reply:
We can improve type inference.

There are a few options.

### Option 1

Use one parameter list and require that `R <: Live[R1]`.

```scala
def withLive[R1, R <: Live[R1], E, A, B](zio: ZIO[R, E, A], f: IO[E, A] => ZIO[R1, E, B]): ZIO[R, E, B] =
```

### Option 2

Create a temporary class to allow incrementally introducing type parameters for each parameter list:

```scala
def withLive[R, E, A](zio: ZIO[R, E, A]): WithLive[R, E, A] = 
  WithLive[R, E, A](zio)

final case class WithLive[R, E, A](zio: ZIO[R, E, A]) {
  def apply[R1, B](f: IO[E, A] => ZIO[R1, E, B]): ZIO[R with Live[R1], E, B] =
    ZIO.environment[R].flatMap(r => live(f(zio.provide(r))))
}
```

### Option 3

Flip order so we can get `R1` before `R`. I don't think this will infer well so I won't right it (well, some parts of it will infer better, but overall worse).